haas/ext/switches/brocade.py

```diff
@@ -0,0 +1,165 @@
+# you may not use this file except in compliance with the
+# License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""A switch driver for the Brocade SDN
+
+Currently the driver uses telnet to connect to the switch's console; in the
+long term we want to be using SNMP.
+"""
+
+import json
+import logging
+import pexpect
+import re
+import requests
+import schema
+
+from sqlalchemy import Column, String, Integer, ForeignKey
+
+from haas.model import Switch
+
+logger = logging.getLogger(__name__)
+
+
+class Brocade(Switch):
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/brocade'
+
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = Column(Integer, ForeignKey('switch.id'), primary_key=True)
+    hostname = Column(String, nullable=False) # Hostname of the SDN
+    username = Column(String, nullable=False) # Username of the SDN
+    password = Column(String, nullable=False) # Password of the SDN
+    node = Column(String, nullable=False) # Name as registered in SDN
+    interface_type = Column(String, nullable=False)
+    dummy_vlan = Column(String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'hostname': basestring,
+            'username': basestring,
+            'password': basestring,
+            'node': basestring,
+            'interface_type': basestring,
+            'dummy_vlan': schema.And(schema.Use(int),
+                                     lambda v: 0 <= v <= 4096,
+                                     schema.Use(str)),
+        }).validate(kwargs)
+
+    def session(self):
+        return self
+
+    def disconnect(self):
+        pass
+
+    def apply_networking(self, action):
+        interface = action.nic.port.label
+        channel = action.channel
+
+        if channel == 'vlan/native':
+            old_native = None
+            old_attachments = filter(lambda a: a.channel == 'vlan/native',
+                                     action.nic.attachments)
+            if len(old_attachments) != 0:
+                old_native = old_attachments[0].network.network_id
+            if action.new_network is None:
+                self._remove_vlan_from_trunk(interface, old_native)
+                self._set_native_vlan(interface, self.dummy_vlan)
+            else:
+                self._set_native_vlan(interface,
+                                      action.new_network.network_id)
+        else:
+            match = re.match(re.compile(r'vlan/(\d+)'), channel)
+            assert match is not None, "HaaS passed an invalid channel to the switch!"
```

Comment:
Agree. The fact that it's an assert for something that can happen without there being a bug in HaaS is ugly, but we should be doing the check. Really, we should check the config on startup too, in which case having an assert here would make perfect since, since it shouldn't be possible then.